File: cafw-backend/app/firewall/middleware.py
import json
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import JSONResponse
from app.database import SessionLocal
from app.firewall.detector import detect_attack, scan_request
from app.firewall.logger import log_attack, is_ip_blocked

WHITELIST_PATHS = [
    "/",
    "/health",
    "/docs",
    "/openapi.json",
    "/redoc",
    "/auth/register",
    "/auth/verify-register",
    "/auth/login",
    "/auth/verify-login",
    "/auth/forgot-password",
    "/auth/reset-password",
    "/auth/check-otp-store",
    "/dashboard/stats",
    "/dashboard/category-breakdown",
    "/dashboard/recent-attacks",
    "/dashboard/top-attackers",
    "/logs/",
    "/logs/count",
    "/rules/",
]

class FirewallMiddleware(BaseHTTPMiddleware):

    async def dispatch(self, request: Request, call_next):

        path = request.url.path

        # Skip whitelisted paths
        if (path in WHITELIST_PATHS
                or path.startswith("/auth/")
                or path.startswith("/docs")
                or path.startswith("/redoc")
                or path.startswith("/openapi")):
            return await call_next(request)

        db = SessionLocal()
        try:
            ip     = request.client.host if request.client else "unknown"
            method = request.method

            # Check 1 — blocked IP
            if is_ip_blocked(db, ip):
                return JSONResponse(
                    status_code=403,
                    content={
                        "error":  "Access denied",
                        "reason": "Your IP has been permanently blocked",
                        "ip":     ip
                    }
                )

            # Check 2 — query params
            query_data = dict(request.query_params)
            attack = scan_request(query_data)
            if attack:
                log_attack(
                    db=db, ip_address=ip, method=method,
                    endpoint=path, category=attack["category"],
                    payload=attack["payload"],
                    rule_matched=attack["matched_pattern"]
                )
                return JSONResponse(
                    status_code=403,
                    content={
                        "error":    "Request blocked",
                        "reason":   attack["category"] + " attack detected",
                        "category": attack["category"]
                    }
                )

            # Check 3 — request body
            if method in ["POST", "PUT", "PATCH"]:
                try:
                    body_bytes = await request.body()
                    if body_bytes:
                        body = json.loads(body_bytes.decode("utf-8"))
                        if isinstance(body, dict):
                            attack = scan_request(body)
                        else:
                            attack = detect_attack(str(body))
                        if attack:
                            log_attack(
                                db=db, ip_address=ip, method=method,
                                endpoint=path, category=attack["category"],
                                payload=attack["payload"],
                                rule_matched=attack["matched_pattern"]
                            )
                            return JSONResponse(
                                status_code=403,
                                content={
                                    "error":    "Request blocked",
                                    "reason":   attack["category"] + " attack detected",
                                    "category": attack["category"]
                                }
                            )
                except Exception:
                    pass

            # Check 4 — URL path itself
            path_attack = detect_attack(path)
            if path_attack:
                log_attack(
                    db=db, ip_address=ip, method=method,
                    endpoint=path, category=path_attack["category"],
                    payload=path_attack["payload"],
                    rule_matched=path_attack["matched_pattern"]
                )
                return JSONResponse(
                    status_code=403,
                    content={
                        "error":    "Request blocked",
                        "reason":   path_attack["category"] + " attack detected",
                        "category": path_attack["category"]
                    }
                )

            return await call_next(request)

        finally:
            db.close()

File: cafw-backend/firewall/middleware.js
const { queryOne } = require('../database');
const { scanRequest, getActivePatterns } = require('./detector');
const { logAttack, logRequest } = require('./logger');

/**
 * Whitelisted paths that bypass firewall inspection.
 */
const WHITELIST_PATHS = [
  '/api/dashboard',
  '/api/logs',
  '/api/rules',
  '/',
  '/health',
];

/**
 * Extract client IP from request.
 */
function getClientIP(req) {
  const forwarded = req.headers['x-forwarded-for'];
  if (forwarded) {
    return forwarded.split(',')[0].trim();
  }
  return req.ip || req.connection?.remoteAddress || '0.0.0.0';
}

/**
 * Express middleware that intercepts all requests and inspects them for attacks.
 */
function firewallMiddleware(req, res, next) {
  // Skip whitelisted paths
  const isWhitelisted = WHITELIST_PATHS.some(path =>
    req.path === path || req.path.startsWith(path + '/')
  );

  if (isWhitelisted) {
    return next();
  }

  const ip = getClientIP(req);

  try {
    // Step 1: Check if IP is blocked
    const blockedIp = queryOne(
      'SELECT * FROM blocked_ips WHERE ip_address = ? AND is_active = 1',
      [ip]
    );

    if (blockedIp) {
      logAttack({
        ip,
        method: req.method,
        endpoint: req.path,
        category: 'Blocked IP',
        payload: '',
        ruleMatched: `IP ${ip} is blocked: ${blockedIp.reason}`,
      });
      logRequest({ ip, method: req.method, endpoint: req.path, blocked: true, category: 'Blocked IP' });
      return res.status(403).json({
        error: 'Blocked by firewall',
        reason: 'Your IP address has been blocked',
        blocked: true,
      });
    }

    // Load active patterns from database (respects rule toggling)
    const activePatterns = getActivePatterns();

    // Step 2: Scan request path
    const pathResult = scanRequest(req.path, activePatterns);
    if (pathResult) {
      logAttack({
        ip,
        method: req.method,
        endpoint: req.path,
        category: pathResult.category,
        payload: pathResult.payload,
        ruleMatched: pathResult.matchedPattern,
      });
      logRequest({ ip, method: req.method, endpoint: req.path, blocked: true, category: pathResult.category });
      return res.status(403).json({
        error: 'Blocked by firewall',
        reason: `Malicious input detected in request path: ${pathResult.category}`,
        blocked: true,
        category: pathResult.category,
      });
    }

    // Step 3: Scan query parameters
    if (req.query && Object.keys(req.query).length > 0) {
      const queryResult = scanRequest(req.query, activePatterns);
      if (queryResult) {
        logAttack({
          ip,
          method: req.method,
          endpoint: req.path,
          category: queryResult.category,
          payload: queryResult.payload,
          ruleMatched: queryResult.matchedPattern,
        });
        logRequest({ ip, method: req.method, endpoint: req.path, blocked: true, category: queryResult.category });
        return res.status(403).json({
          error: 'Blocked by firewall',
          reason: `Malicious input detected in query parameters: ${queryResult.category}`,
          blocked: true,
          category: queryResult.category,
        });
      }
    }

    // Step 4: Scan request body
    if (req.body && typeof req.body === 'object' && Object.keys(req.body).length > 0) {
      const bodyResult = scanRequest(req.body, activePatterns);
      if (bodyResult) {
        logAttack({
          ip,
          method: req.method,
          endpoint: req.path,
          category: bodyResult.category,
          payload: bodyResult.payload,
          ruleMatched: bodyResult.matchedPattern,
        });
        logRequest({ ip, method: req.method, endpoint: req.path, blocked: true, category: bodyResult.category });
        return res.status(403).json({
          error: 'Blocked by firewall',
          reason: `Malicious input detected in request body: ${bodyResult.category}`,
          blocked: true,
          category: bodyResult.category,
        });
      }
    }

    // Request is clean - log and pass through
    logRequest({ ip, method: req.method, endpoint: req.path, blocked: false, category: null });
    next();
  } catch (err) {
    console.error('Firewall middleware error:', err.message);
    next();
  }
}

module.exports = { firewallMiddleware, getClientIP };
